### PR TITLE
Roles

### PR DESCRIPTION
**Pull request by tyler-tran-goldenowl**
# Description
- Add roles to user (currently, there are 3 roles: admin, buyer, seller)
- Add model Shop
- Add associations between User, Shop, Product
- Add UIs for Shop#new and Shop#show
- Add rspec tests for models and requests

# Screenshots 
### Shop#new 
Shop#new
### Shop#show
Shop#show

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -1,7 +1,15 @@
...
-  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+
   root "home#index"
+
+  resources :users do
+    resource :shop
+  end
   
-  resources :products
+  resources :shop do
```
Comment (nathan-phan-goldenowl) [inline config/routes.rb:10]: Should be `resources :shops`

Comment (tathai2610) [reply config/routes.rb:10]: > Should be `resources :shops`

noted and updated.

File: app/controllers/shops_controller.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  def new 
+    @shop = Shop.new
+  end
+
+  def create 
+    @shop = Shop.new(shop_params.merge(user: @user))
+    if @shop.save
+      @user.add_role :seller 
+      flash[:success] = "Congratulations! You have successfuly open your own shop on Planty!"
+      redirect_to user_shop_path(@user), method: :get
```
[the remark this answers is not in the document]
Comment (tathai2610) [reply app/controllers/shops_controller.rb:21]: > `redirect_to user_shop_path(@user)` is enough, no need `method` params behind

noted and updated.

File: app/controllers/products_controller.rb [hunk 3]
```diff
@@ -2,19 +2,21 @@ class ProductsController < ApplicationController
...
   
-  def show 
-    
+  def show
   end
 
   private 
 
   def set_product
+    @shop = Shop.find(params[:shop_id])
```
[the remark this answers is not in the document]
Comment (tathai2610) [reply app/controllers/products_controller.rb:19]: > this should be extract to another method, for ex: `set_shop`

noted and updated.

File: app/controllers/products_controller.rb [hunk 4]
```diff
@@ -2,19 +2,21 @@ class ProductsController < ApplicationController
...
-    @products = Product.all
-    @products = Product.in_category(params[:category]) if params[:category]
-    @pagy, @products = pagy(@products, items: 12)
-    @categories = Category.all.order(:name)
+    unless params[:shop_id]
+      @products = Product.all
+      @products = Product.in_category(params[:category]) if params[:category]
+      @pagy, @products = pagy(@products, items: 12)
+      @categories = Category.all.order(:name)
+    end
```
[the remark this answers is not in the document]
Comment (tathai2610) [reply app/controllers/products_controller.rb:10]: > and what if there is params[:shop_id]?

Currently, I don't use the Product#index path which provides params[:shop_id], so I only have the code for 1 situation. The other condition will be updated if I use the other path. Or should I include both cases?

[the remark this answers is not in the document]
Comment (nathan-phan-goldenowl) [reply app/controllers/products_controller.rb:10]: in which case params[:shop_id] will be passed in Product#index?

[the remark this answers is not in the document]
Comment (tathai2610) [reply app/controllers/products_controller.rb:10]: > in which case params[:shop_id] will be passed in Product#index?

@nathan-phan-goldenowl `shop_products_path` will pass in `params[:shop_id]`. It is meant to show all products of a shop, but I already included this view in `user_shop_path`, when showing the shop interface

[the remark this answers is not in the document]
Comment (nathan-phan-goldenowl) [reply app/controllers/products_controller.rb:10]: so the it mean you won't use params[:shop_id] in product#index in the future, right?

[the remark this answers is not in the document]
Comment (tathai2610) [reply app/controllers/products_controller.rb:10]: > so the it mean you won't use params[:shop_id] in product#index in the future, right?

@nathan-phan-goldenowl it is very likely as you said

**Review by nathan-phan-goldenowl — APPROVED** [review]

**Pull request merged**